nomad/structs/acl.go

```diff
@@ -360,6 +368,12 @@ func (a *ACLRole) Copy() *ACLRole {
 // roles.
 type ACLRolesUpsertRequest struct {
 	ACLRoles []*ACLRole
+
+	// AllowMissingPolicies skips the ACL Role policy link verification and is
+	// used by the replication process. The replication cannot ensure policies
+	// are present before ACL Roles are replicated.
+	AllowMissingPolicies bool
```

Comment:
Yeah I can't see any way to avoid this because blocking queries can't be made across a multiple tables. But there's nothing in this PR that changes how queries of `ACLRoles` work. How do we handle a role with a missing policy when the role gets checked?

Reply:
When you say checked, is this referring to a token that links to an ACL role which is being used and resolved for API authorisation? If so, this will work in the same way as an ACL token which has policy linkings where the token has been replicated but the policy has not yet. The role policy links are unpacked and the policy is read from state; if the policy is not found within state, then the permissions are not applied to the token and we continue onto the next policy in the list: https://github.com/hashicorp/nomad/blob/f-gh-13120-sso-umbrella/nomad/acl.go#L152-L166

In a case where a role has been replicated, but the single policy referenced has not, use of the token will result in a permission denied error, until the policy is also replicated. Generally this time gap should be very small as all the replication functions use blocking queries.

Reply:
Aye, that all makes sense. Most of the time that'll mean briefly not getting access to something you should in a non-authoritative region, but that's expected given the replication is async. 

I wasn't able to figure out any scenario where this can cause "unexpected open" behavior expect in cases where policies were updated in separate transactions in the authoritative region. But in that case (a) the "unexpectedly open" behavior will _also_ happen on the authoritative region and it's a user error, so replication just widens the window, and (b) that scenario is also what we already have today with tokens. So :+1: